### ISSUE TITLE
Anonymous access to SMB with the client login

### ISSUE THREAD

**Issue by SimonGurney** [TalAloni/SMBLibrary#205]
Hi,

We use this library for scannign for SMB shares and enumerating files, but we cant find any way to check shares which have anonymous browsing.

For instance, if we have an SMB server which allows anonymous access then we can view all the shares in Windows file explorer but the ISMBClient Login function fails.

Is there an anonymous login option for the client connection?

**Comment by TalAloni** [comment]
Yes,
Username: Guest
Password: "" (empty string)

**Comment by SimonGurney** [comment]
Hmmm, that doesn't work.

It's strange.  The windows native explorer can browse them fine but using this library it's doesn't work at all.  The login call just fails.  Should I not be trying to call Login?

**Comment by SimonGurney** [comment]
Pcapped it.

Windows is sending empty username and empty NTLM response.

If I send an empty username and password with this library it fails to login.

**Comment by TalAloni** [comment]
There is the NegotiateFlags.Anonymous which we do not set, I wonder if setting this flag will resolve the issue, can you check?
(Should be done in
NTLMAuthenticationHelper.GetNegotiateMessage
and 
NTLMAuthenticationHelper.GetAuthenticateMessage

**Comment by alexbrozych** [comment]
Hi,

I used Wireshark to compare what this library is sending versus what Windows Explorer sends to be able to list shares without auth, this is what Windows Explorer is sending, and found that Windows Explorer sends Empty Lan Manager and NTLM Response, as seen here
![Screenshot 2023-07-27 155457](https://github.com/TalAloni/SMBLibrary/assets/120374931/b156f175-8bdc-42e2-aeb8-4c91e658c42e)

When using this library and logging in with blank username and password, it sends this
![Screenshot 2023-07-27 155516](https://github.com/TalAloni/SMBLibrary/assets/120374931/22df15d3-4c80-42ac-88b5-4362c097eaa9)

Additionally, I set the NegotiateFlags.Anonymous within those two functions, and it does not seem to have changed the behaviour. Wireshark capture shows the Anonymous flag as being properly set, however the server is still returning STATUS_LOGON_FAILURE

I also checked the behaviour of Windows Explorer, and it is not setting the Anonymous flag for NTMLSSP_NEGOTIATE, but is for NTMLSSP_AUTH. Even matching this behaviour, server still responds with STATUS_LOGON_FAILURE

**Comment by TalAloni** [comment]
Thanks for the detailed information.
I'll see if I can set up a server to behave this way and produce the desirable NTLM messages.
I have some higher priority items, so will not get to it soon though.

**Comment by TalAloni** [comment]
I've pushed a possible fix to this, it will no login as anonymous when username == string.Empty and password == string.Empty (per https://msdn.microsoft.com/en-us/library/cc236700.aspx)

Please test and let me know if this works for you

**Comment by SimonGurney** [comment]
Oh brilliant, we'll test this tomorrow :)

**Comment by alexbrozych** [comment]
Tested these changes, getting Exception within GenerateSigningKey when calling Login with empty username and password due to null session key being returned from GetAuthenticateMessage
![image](https://github.com/TalAloni/SMBLibrary/assets/120374931/bb34502f-e580-44bb-bbe1-0d4382b57658)
Solution appears to be adding these five lines from NTMLv2 auth codepath to the empty username / password block such that it becomes
```csharp
if (userName == String.Empty && password == String.Empty)
{
    // https://msdn.microsoft.com/en-us/library/cc236700.aspx
    authenticateMessage.LmChallengeResponse = new byte[1];
    authenticateMessage.NtChallengeResponse = new byte[0];
    NTLMv2ClientChallenge clientChallengeStructure = new NTLMv2ClientChallenge(time, clientChallenge, challengeMessage.TargetInfo, spn);
    byte[] clientChallengeStructurePadded = clientChallengeStructure.GetBytesPadded();
    byte[] ntProofStr = NTLMCryptography.ComputeNTLMv2Proof(challengeMessage.ServerChallenge, clientChallengeStructurePadded, password, userName, domainName);
    byte[] responseKeyNT = NTLMCryptography.NTOWFv2(password, userName, domainName);
    sessionBaseKey = new HMACMD5(responseKeyNT).ComputeHash(ntProofStr);
    keyExchangeKey = sessionBaseKey;
}
```

**Comment by TalAloni** [comment]
Thanks for the feedback.
I did something slightly different to also accommodate NTLM v1, please test and let me know if all is well now.

**Comment by alexbrozych** [comment]
![image](https://github.com/TalAloni/SMBLibrary/assets/120374931/7e19fd67-a54d-4ae8-a0b2-2d8087a566d2)
null ntProofStr value causes call to ComputeHash to fail, changing ntProofStr to be initially set to `new byte[0]` instead of `null` seems to fix the issue for me without seeming to affect case where username and password are provided.

**Comment by TalAloni** [comment]
My mistake. corrected now. please test and update.

**Comment by alexbrozych** [comment]
Yes, these latest changes work fine. Both anonymous and authenticated login are working for me.

**Comment by SimonGurney** [comment]
We install from nuget, do you know when the next release is?

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by TalAloni** [comment]
When a bugfix or a useful feature will be merged.
Supporting anonymous users is a niche feature that does not warrant a release, sorry.

**Comment by SimonGurney** [comment]
Ah this is definitely a super useful feature.  So many systems host anonymous shares, everything from printers and routers all the way to enterprise storage (our use case).  

At the moment there is nothing available in .net to talk to them, and yet it works out the box on Windows.

If you do have a change of heart, could you bump this ticket?  In the mean time we'll try and build from the git red :)

Thanks for sorting!

**Comment by TalAloni** [comment]
OK, I've released v1.5.0.1 including this change.
